aiohttp/web_response.py

```diff
@@ -662,7 +673,12 @@ async def write_eof(self, data: bytes = b"") -> None:
             await super().write_eof()
 
     async def _start(self, request: "BaseRequest") -> AbstractStreamWriter:
-        if not self._chunked and hdrs.CONTENT_LENGTH not in self._headers:
+        if self._must_be_empty_body and should_remove_content_length(
+            request.method, self.status
+        ):
+            if hdrs.CONTENT_LENGTH in self._headers:
+                del self._headers[hdrs.CONTENT_LENGTH]
+        elif not self._chunked and hdrs.CONTENT_LENGTH not in self._headers:
```

Comment:
I think its ok to run on HEAD

> A server MAY send a Content-Length header field in a response to a HEAD request ([Section 9.3.2](https://www.rfc-editor.org/rfc/rfc9110.html#HEAD)); a server MUST NOT send Content-Length in such a response unless its field value equals the decimal number of octets that would have been sent in the content of a response if the same request had used the GET method.

Reply:
https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6-7

Reply:
Hmm, I think if the HEAD request was handled automatically by aiohttp, then it will be fine, as it just treats it as a GET request and will have included the entire body of the response (which will then be stripped later). We should ensure there is a test that verifies the HEAD request doesn't receive a 0 Content-Length.

But, if a user manually handles a HEAD request, then this behaviour is incorrect, as they may return an empty body for the HEAD request, but not for a GET request. I don't know if we have an easy way to tell if this is from a user handler or an automatic aiohttp one...

Reply:
> Hmm, I think if the HEAD request was handled automatically by aiohttp, then it will be fine, as it just treats it as a GET request and will have included the entire body of the response (which will then be stripped later). We should ensure there is a test that verifies the HEAD request doesn't receive a 0 Content-Length for this case.

I think this is covered by `test_head_response_keeps_content_length_of_original_body`
> 
> But, if a user manually handles a HEAD request, then this behaviour is incorrect, as they may return an empty body for the HEAD request, but not for a GET request. I don't know if we have an easy way to tell if this is from a user handler or an automatic aiohttp one...

Need to think about how to handle this case

Reply:
> I think this is covered by `test_head_response_keeps_content_length_of_original_body`

Yes, and I can see in test_web_functional.py we have a HEAD test there: test_head_returns_empty_body
Would be ideal to just add 1 more assert to that test for the Content-Length too.

Reply:
> I think its ok for 304 as well

I think as per the above discussion, we can't do this for 304, as it will always be a user-supplied response. So, it should skip that step in all cases of 304, I believe.

Reply:
The user can obviously set Content-Length themselves if they want to include that information.

Reply:
adjusted `test_head_returns_empty_body` 02f22fa3626fd1c1e80003845902d1c38548837f

Reply:
> I think as per the above discussion, we can't do this for 304, as it will always be a user-supplied response. So, it should skip that step in all cases of 304, I believe.

I should probably do that in `_prepare_headers` instead... will adjust

Reply:
oh, its already done there

Reply:
`test_rm_content_length_1xx_204_304_responses` covers the 304

Reply:
> oh, its already done there

That's because the helper function is still checking 304, right? The spec treats it the same as HEAD requests. So, basically don't delete the header if the user set it explicitly, but also don't add the header automatically if they didn't.

I guess the HEAD request case, we can go with the check of whether the body is set, as you've done (but, that does still leave a small chance for issues if the user passes a body by mistake, but that's probably the best we'll get for now).

Reply:
I had a quick look to see if we could get back to the handlers/urldispatcher (as we could then check if the same handler is registered for GET/HEAD), but looks like there's no path to it. So, will stick with what you've done.